### PR TITLE
[benqprojector] Add discovery service

### PR DESCRIPTION
**Pull request by mlobstein**
Add discovery service to automatically locate BenQ projectors on the local network.

### CODE REVIEW

File: bundles/org.openhab.binding.benqprojector/src/main/java/org/openhab/binding/benqprojector/internal/discovery/MulticastListener.java [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+        if (beacon.toUpperCase(Locale.ENGLISH).contains("BENQ") && beacon.contains("VideoProjector")) {
+            String[] parameterList = beacon.replace(">", "").split("<-");
+
+            for (String parameter : parameterList) {
+                String[] keyValue = parameter.split("=");
+
+                if (keyValue.length == 2 && keyValue[0].contains("UUID") && !keyValue[1].isEmpty()) {
+                    Map<String, Object> properties = new HashMap<>();
+                    properties.put(THING_PROPERTY_MAC, keyValue[1]);
```
Comment (jlaur) [inline bundles/org.openhab.binding.benqprojector/src/main/java/org/openhab/binding/benqprojector/internal/discovery/MulticastListener.java:123]: Here you could use framework constant `Thing.PROPERTY_MAC_ADDRESS`.

**Review by jlaur — CHANGES_REQUESTED** [review]
Thanks for this enhancement. Only some very minor comments added.

File: bundles/org.openhab.binding.benqprojector/src/main/java/org/openhab/binding/benqprojector/internal/discovery/BenqProjectorDiscoveryService.java [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+                    String uid = (String) thingProperties.get(THING_PROPERTY_MAC);
+                    String ipAddress = (String) thingProperties.get(THING_PROPERTY_HOST);
+
+                    if (uid != null) {
+                        logger.trace("Projector with UID {} discovered at IP: {}", uid, ipAddress);
+
+                        ThingUID thingUid = new ThingUID(THING_TYPE_PROJECTOR_TCP, uid);
+                        logger.trace("Creating BenQ projector discovery result for: {}, IP={}", uid, ipAddress);
+                        thingDiscovered(DiscoveryResultBuilder.create(thingUid).withProperties(thingProperties)
+                                .withLabel("BenQ Projector " + uid).withRepresentationProperty(THING_PROPERTY_MAC)
```
[the remark this answers is not in the document]
Comment (mlobstein) [reply bundles/org.openhab.binding.benqprojector/src/main/java/org/openhab/binding/benqprojector/internal/discovery/BenqProjectorDiscoveryService.java]: Done, thanks!

File: bundles/org.openhab.binding.benqprojector/src/main/java/org/openhab/binding/benqprojector/internal/discovery/BenqProjectorDiscoveryService.java [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+
+    private synchronized void discover() {
+        logger.debug("Discovery job is running");
+        MulticastListener benqMulticastListener;
+        String local = "127.0.0.1";
+
+        try {
+            String ip = networkAddressService.getPrimaryIpv4HostAddress();
+            benqMulticastListener = new MulticastListener((ip != null ? ip : local));
+        } catch (SocketException se) {
```
Comment (jlaur) [inline bundles/org.openhab.binding.benqprojector/src/main/java/org/openhab/binding/benqprojector/internal/discovery/BenqProjectorDiscoveryService.java:121]: `SocketException` is a subclass of `IOException`, so is there any need to catch this explicitly when handled exactly the same way as below? Only difference is the logged text "Socket" vs. "IO", but the message should be sufficient.

File: bundles/org.openhab.binding.benqprojector/README.md [hunk 4]
```diff
@@ -11,7 +11,8 @@ This binding supports two thing types based on the connection used: `projector-s
 
 ## Discovery
 
-The projector thing cannot be auto-discovered, it has to be configured manually.
+If the projector has a built-in ethernet port connected to the same network as the openHAB server and supports AMX Device Discovery, the thing will be discovered automatically.
```
Comment (jlaur) [inline bundles/org.openhab.binding.benqprojector/README.md:14]:
```suggestion
If the projector has a built-in Ethernet port connected to the same network as the openHAB server and supports AMX Device Discovery, the thing will be discovered automatically.
```

**Review by jlaur — CHANGES_REQUESTED** [review]
Sorry, two minor additional findings on last walk-through.

File: bundles/org.openhab.binding.benqprojector/src/main/java/org/openhab/binding/benqprojector/internal/discovery/BenqProjectorDiscoveryService.java (hunk 3 above)
Comment (mlobstein) [reply bundles/org.openhab.binding.benqprojector/src/main/java/org/openhab/binding/benqprojector/internal/discovery/BenqProjectorDiscoveryService.java]: Removed the extra catch block.

File: bundles/org.openhab.binding.benqprojector/README.md (hunk 4 above)
Comment (mlobstein) [reply bundles/org.openhab.binding.benqprojector/README.md]: fixed (2 places), thanks!

**Comment by jlaur** [comment]
@mlobstein - spotless check fails because of unused import, SocketException. After fixing that we are ready to go.

**Review by jlaur — APPROVED** [review]
LGTM

**Pull request merged**